v3/src/components/map/models/map-grid-model.ts

```diff
@@ -0,0 +1,205 @@
+        boundsSouth = latLongBounds?.getSouth() ?? 0,
+        boundsNorth = latLongBounds?.getNorth() ?? 0
+
+      if (boundsWest === boundsEast) {
+        boundsWest -= 5 // arbitrary offset
+        boundsEast += 5
+      }
+      if (boundsNorth === boundsSouth) {
+        boundsNorth += 5 // arbitrary offset
+        boundsSouth -= 5
+      }
+
+      self.setGridWidth(Math.min((boundsEast - boundsWest) /
+        kGridCellCount, (boundsNorth - boundsSouth) / kGridCellCount))
+
+      const gridSize = self.gridSize,
+        numToWest = Math.ceil((refLong - boundsWest) / gridSize)
+      startWest = refLong - gridSize * numToWest
+
+      const numToSouth = Math.ceil((refLat - boundsSouth) / gridSize)
+      startSouth = refLat - gridSize * numToSouth
+
+      for (let longIndex = 0; longIndex < 2 * numToWest; longIndex++) {
+        for (let latIndex = 0; latIndex < 2 * numToSouth; latIndex++) {
+          rectArray.setRect(longIndex, latIndex, new RectRecord(new LatLngBounds([
+            [startSouth + (latIndex + 1) * gridSize, startWest + longIndex * gridSize],
+            [startSouth + latIndex * gridSize, startWest + (longIndex + 1) * gridSize]
+          ])))
+        }
+      }
+    }
+
+    const zeroCounts = () => {
+      self.rectArray.forEachRect((rect) => {
+        rect.count = 0
+        rect.cases = []
+      })
+      self.rectArray.maxCount = 0
+    }
+
+    const computeCounts = () => {
+      if (self.dataConfiguration) {
+        const {
+            dataConfiguration: {dataset, caseDataArray},
+            rectArray, gridSize
+          } = self,
+          dataConfiguration = self.dataConfiguration
+
+        zeroCounts()
+        caseDataArray.forEach(({caseID}: { plotNum: number, caseID: string }) => {
+          const long = dataset?.getNumeric(caseID, dataConfiguration.attributeID('long')),
+            lat = dataset?.getNumeric(caseID, dataConfiguration.attributeID('lat'))
+          if (long !== undefined && lat !== undefined) {
+            const longIndex = Math.floor((long - startWest) / gridSize),
+              latIndex = Math.floor((lat - startSouth) / gridSize)
+            rectArray.addCaseToRect(longIndex, latIndex, caseID)
+          }
+        })
+      }
+    }
+
+    return {
+      actions: {
+        initializeRectArray() {
+          const cases = self.dataConfiguration?.caseDataArray,
+            bounds = self.dataConfiguration ? getLatLongBounds(self.dataConfiguration) : undefined
+          if (cases && cases.length > 0 && bounds && bounds.isValid()) {
+            setupRectangles()
+            computeCounts()
+            self.rectArray.deleteZeroRects()
+          }
+        },
+        clear() {
+          self.rectArray.clear()
+        },
+        updateSelection() {
+          self.forEachRect((rect) => {
+            rect.selected = rect.cases.every((aCase) => self.dataConfiguration?.dataset?.isCaseSelected(aCase))
+          })
+        },
+      }
+    }
```

Comment:
The mix of responsibilities between the `MapGridModel` and the `RectArray` seems unnecessarily complicated. Couldn't most of the functions defined here be methods of the `RectArray` class (which I've elsewhere suggested could be called something like `LatLngGrid`) and then, for instance, the maintenance of the `maxCount` property wouldn't be as reliant on whether the caller calls the class methods in a particular order?